quantization: deprecate observer compute_dtype and replace with is_dynamic

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85431

Summary:

This PR deprecates the `compute_dtype` field on observers, and replaces
it with the `is_dynamic` field on observers.  This is better aligned
with the reference model spec.

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @jgong5 @Xia-Weiwen @leslie-fang-intel @ezyang @SherlockNoMad @soumith @EikanWang @wenzhe-nrv

vkuzo committed:
quantization: deprecate observer compute_dtype and replace with is_dynamic

Summary:

This PR deprecates the `compute_dtype` field on observers, and replaces
it with the `is_dynamic` field on observers.  This is better aligned
with the reference model spec.

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

[ghstack-poisoned]

vkuzo committed:
Update on "quantization: deprecate observer compute_dtype and replace with is_dynamic"


Summary:

This PR deprecates the `compute_dtype` field on observers, and replaces
it with the `is_dynamic` field on observers.  This is better aligned
with the reference model spec.

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/utils.py:
```diff
@@ -229,41 +229,42 @@ def op_is_int8_dynamically_quantized(qconfig) -> bool:
     static_dtypes = [torch.quint8, torch.qint8, torch.quint4x2]
     if weight.dtype in static_dtypes:
-        if hasattr(activation, 'compute_dtype') and activation.compute_dtype in static_dtypes:
+        # if hasattr(activation, 'compute_dtype') and activation.compute_dtype in static_dtypes:
```
remove?

jerryzh168 reviewed (approved):
great, thanks a lot for the clean up!

vkuzo committed:
Update on "quantization: deprecate observer compute_dtype and replace with is_dynamic"


Summary:

This PR deprecates the `compute_dtype` field on observers, and replaces
it with the `is_dynamic` field on observers.  This is better aligned
with the reference model spec.

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -1,4 +1,4 @@
-from typing import Any, Dict, List, Optional, Set, Tuple, Union, Type
+from typing import Any, Dict, List, Optional, Set, Tuple, Union, Type, Callable
```
There were various mypy errors in the master branch version of this file in my local env, so this PR also fixes these errors.  Technically some of this stuff could be in a separate PR, lmk if that's needed.

jerryzh168 commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -131,11 +131,13 @@ def _replace_observer_with_quantize_dequantize_node_decomposed(
-        compute_dtype = activation_post_process.compute_dtype  # type: ignore[attr-defined]
+
+    is_dynamic = False
+    if hasattr(activation_post_process, "is_dynamic"):
```
nit: should we make this a required attribute going forward?

jerryzh168 commented on torch/ao/quantization/fx/convert.py:
```diff
@@ -131,11 +131,13 @@ def _replace_observer_with_quantize_dequantize_node_decomposed(
+        is_dynamic = activation_post_process.is_dynamic  # type: ignore[assignment]
+
     if dtype in [torch.quint8, torch.qint8, torch.qint32] and \
-            not hasattr(activation_post_process, 'compute_dtype'):
+            (not is_dynamic):
```
nit (can be done in a separate PR): I think the overall checking can be made clearer as well for this code, right now it is:
```
if dtype in [torch.quint8, torch.qint8, torch.qint32] and not is_dynamic:
   ...
elif is_dynamic:
   ...
else:
   # fp16
```

I think a clearer one should be:
```
if is_dynamic:
   ...
else:
   if dtype in [torch.quint8, torch.qint8, torch.qint32]:
      ...
    else:
      # fp16
      ...
```

jerryzh168 replied:
also we might want to add dtype check for is_dynamic branch as well, looks like we only support uint8, int8, fp16 in the non-decomposed function and uint8 and int8 in the decomposed function

jerryzh168 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -190,12 +190,25 @@ def get_quantize_node_info(
-    if dtype in [torch.quint8, torch.qint8, torch.qint32] and \
-            not hasattr(activation_post_process, 'compute_dtype'):
+
+    if is_dynamic:
```
I think this function is removed already? could you rebase on master?

jerryzh168 replied:
oh it's not landed yet..

jerryzh168 reviewed (approved):
looks good, thanks a lot, please wait until https://github.com/pytorch/pytorch/pull/89272 is merged so we can remove the changes to `get_quantize_info` function, also had some optional nits that can be fixed later

vkuzo committed:
Update on "quantization: deprecate observer compute_dtype and replace with is_dynamic"


Summary:

This PR deprecates the `compute_dtype` field on observers, and replaces
it with the `is_dynamic` field on observers.  This is better aligned
with the reference model spec.

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

cc jerryzh168 jianyuh raghuramank100 jamesr66a jgong5 Xia-Weiwen leslie-fang-intel ezyang SherlockNoMad soumith EikanWang wenzhe-nrv

[ghstack-poisoned]

vkuzo committed:
Update on "quantization: deprecate observer compute_dtype and replace with is_dynamic"


Summary:

This PR deprecates the `compute_dtype` field on observers, and replaces
it with the `is_dynamic` field on observers.  This is better aligned
with the reference model spec.

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

cc jerryzh168 jianyuh raghuramank100 jamesr66a jgong5 Xia-Weiwen leslie-fang-intel ezyang SherlockNoMad soumith EikanWang wenzhe-nrv

[ghstack-poisoned]

vkuzo committed:
Update on "quantization: deprecate observer compute_dtype and replace with is_dynamic"


Summary:

This PR deprecates the `compute_dtype` field on observers, and replaces
it with the `is_dynamic` field on observers.  This is better aligned
with the reference model spec.

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

cc jerryzh168 jianyuh raghuramank100 jamesr66a jgong5 Xia-Weiwen leslie-fang-intel ezyang SherlockNoMad soumith EikanWang wenzhe-nrv

[ghstack-poisoned]

vkuzo commented:
@pytorchbot merge
